WordPress/Classes/Services/SuggestionService.m

```diff
@@ -91,23 +82,84 @@ - (void)updateSuggestionsForSiteID:(NSNumber *)siteID
+    NSArray *sites = [context executeFetchRequest:request error:&error];
+    if (error) {
+        DDLogError(@"Could not fetch Autocompleter for site %@, error", siteID, error);
+        return nil;
+    } else if (sites.count > 1) {
+        DDLogError(@"Retrieved more than one Autocompleter for site %@", siteID);
+        return nil;
+    } else {
+        return [sites firstObject];
+    }
+}
+
+- (Autocompleter *)insertAutocompleterForSiteID:(NSNumber *)siteID
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    Autocompleter *autocompleter = [NSEntityDescription insertNewObjectForEntityForName:@"Autocompleter" inManagedObjectContext:context];
+    autocompleter.siteID = siteID;
+    return autocompleter;
+}
+
+- (void)deleteUserAutocompletesForAutocompleter:(Autocompleter *)autocompleter
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    for (UserAutocomplete *userAutocomplete in autocompleter.userAutocompletes) {
+        [context deleteObject:[context objectWithID:userAutocomplete.objectID]];
+    }
+}
+
+- (void)updateAndSaveUserAutocompletesForSiteID:(NSNumber *)siteID suggestions:(NSArray *)suggestions
+{
+    Autocompleter *autocompleter = [self retrieveAutocompleterForSiteID:siteID];
+
+    if (autocompleter) {
+        [self deleteUserAutocompletesForAutocompleter:autocompleter];
+    } else {
+        autocompleter = [self insertAutocompleterForSiteID:siteID];
+    }
+
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    for (NSDictionary *suggestion in suggestions) {
+        UserAutocomplete *userAutocomplete = [NSEntityDescription insertNewObjectForEntityForName:@"UserAutocomplete" inManagedObjectContext:context];
+        userAutocomplete.username = [suggestion stringForKey:@"user_login"];
+        userAutocomplete.displayName = [suggestion stringForKey:@"display_name"];
+        userAutocomplete.imageURL = [NSURL URLWithString:[suggestion stringForKey:@"image_URL"]];
+        userAutocomplete.autocompleter = autocompleter;
+    }
+
+    NSError *error;
+    if (![context save:&error]) {
+        DDLogError(@"Can't save UserAutocompletes for site %@, error: %@", siteID, error);
+    } else {
+        // send the siteID with the notification so it could be filtered out
+        [[NSNotificationCenter defaultCenter] postNotificationName:SuggestionListUpdatedNotification object:siteID];
+    }
+}
+
 - (BOOL)shouldShowSuggestionsForSiteID:(NSNumber *)siteID
 {
     if (!siteID) {
         return NO;
     }
-    
-    WordPressAppDelegate *appDelegate = [WordPressAppDelegate shared];
-    
-    NSArray *suggestions = [self.suggestionsCache objectForKey:siteID];
-    
-    // if the device is offline and suggestion list is not yet retrieved
-    if (!appDelegate.connectionAvailable && !suggestions) {
+
+    // if the device is offline
+    if (![WordPressAppDelegate shared].connectionAvailable) {
```

Comment:
If the device is offline can't we just rely on the cached data? I guess it worked like that before. Or am I missing something?

Reply:
Hmm, I was comparing with the previous logic, it was:

```
     if (!appDelegate.connectionAvailable && !suggestions) { 
```

The current logic doesn't check for existing suggestions, so it wouldn't show it while offline, right?

Reply:
True! I should re-enable the previous logic for suggestions in Reader to keep things like they were before. I might disable offline support for @-mentions in Gutenberg though, since that will be handled separately in a later PR (with extra UI to indicate to the user that the list of @-mentions is not fresh).

Reply:
Marking this as resolved because the latest commits check this condition properly, so this is fixed.